### PR TITLE
fix: Anthropic header handling — 6 bugs (Issue #35)

### LINKED ISSUE #35: fix: anthropic-beta header usa global upstream_type en vez de per-route
# N3: anthropic-beta header usa global upstream_type en vez de per-route

**Severidad:** MEDIUM
**Origen:** CodeRabbit nitpick from PR#31
**Archivo(s):** src/proxy/retry.rs, src/config.rs
**Estado:** Pendiente
**Fecha detección:** 2026-04-28

## Descripción del Problema

En `src/proxy/retry.rs` líneas 52-53 y 224-227, el header `anthropic-beta` es condicionado por `config.get_upstream_type(upstream_name)` que retorna `self.upstream_type` (un valor global de configuración), ignorando completamente el parámetro `upstream_name` que se le pasa.

**Root cause:** En `config.rs` líneas 573-575, `get_upstream_type()` tiene la firma para aceptar `upstream_name: &str` pero ignora el parámetro:

```rust
pub fn get_upstream_type(&self, _upstream_name: &str) -> &str {
    &self.upstream_type  // Retorna global, ignora parámetro
}
```

**Impacto:** Cuando un modelo es mapeado a un upstream Anthropic vía model_map, pero el upstream_type global es "nim", el header `anthropic-beta` NO se envía. Esto significa que características específicas de Anthropic (como extended thinking) silenciosamente fallan para modelos ruteados.

## Análisis de Causa Raíz

El sistema de configuración fue diseñado para soportar ruteo por model_map (diferentes upstreams para diferentes modelos), pero la implementación de `get_upstream_type()` fue escrita como un stub que solo retorna el valor global. Esto crea una inconsistencia:

1. La model_map puede rutear `claude-sonnet-4-6` a un upstream Anthropic
2. Pero el config global tiene `upstream_type = "nim"` (para otros modelos)
3. Cuando se evalúa el header anthropic-beta, se usa el valor global "nim"
4. El header no se añade → thinking extendido no funciona → degradación silenciosa

El código afectado en retry.rs:
```rust
// ❌ Problema: get_upstream_type ignora el upstream_name pasado
let upstream_type = config.get_upstream_type(upstream_name);  // Siempre "nim" si global es nim
if upstream_type == "anthropic" {
    headers.insert("anthropic-beta", ...);
}
```

## Código Afectado

**Archivo:** `src/proxy/retry.rs`
- Líneas 52-53: Primer uso del header anthropic-beta
- Líneas 224-227: Segundo uso del header anthropic-beta

**Archivo:** `src/config.rs`
- Líneas 573-575: Implementación de `get_upstream_type()` que ignora el parámetro

### Solución Simple (Recomendada para hotfix rápido)

En `retry.rs`, verificar directamente si `upstream_name` contiene "anthropic" o matchea un patrón conocido de upstream Anthropic, en lugar de confiar en el config global:

```rust
// src/proxy/retry.rs

fn prepare_headers(config: &Config, upstream_name: &str) -> HeaderMap {
    let mut headers = HeaderMap::new();

    // ... otros headers

    // ✅ Solución simple: verificar por nombre del upstream
    let is_anthropic_upstream = upstream_name.contains("anthropic");
    if is_anthropic_upstream {
        headers.insert(
            "anthropic-beta",
            HeaderValue::from_static("prompt-caching-2024-07-31"),
        );
    }

    headers
}
```

### Solución Propia (Requiere cambios de schema)

Fix `get_upstream_type()` para que realmente haga lookup por upstream_name desde la model_map, usando el tipo en retry.rs:

```rust
// src/config.rs

pub fn get_upstream_type(&self, upstream_name: &str) -> &str {
    // Buscar en model_map primero
    if let Some(Some(model_entry)) = self.model_map.get(upstream_name) {
        // La model_map necesitaria un campo upstream_type opcional
        if let Some(ref specific_type) = model_entry.upstream_type {
            return specific_type;
        }
    }

    // Default al global
    &self.upstream_type
}
```

Esto requiere añadir un campo `upstream_type` opcional a las entradas de model_map:

```rust
// src/config.rs - ModelMap entry
pub struct ModelMapEntry {
    pub upstream_url: String,
    pub upstream_model: String,
    pub upstream_type: Option<String>, // ← Nuevo campo
}
```

Con variables de entorno:
```bash
# Antes: MODEL_MAP_claude_sonnet_4_6=https://api.anthropic.com/v1:claude-3-5-sonnet-20241022
# Despues: MODEL_MAP_claude_sonnet_4_6=https://api.anthropic.com/v1:claude-3-5-sonnet-20241022:anthropic
```

## Riesgos y Consideraciones

1. **Simple fix risks**: Hardcodea la convención de nombres ("anthropic" en el nombre). Si alguien nombra su upstream "anthropic-proxy" pero es realmente OpenRouter, el header incorrecto se añade.

2. **Proper fix risks**: Requiere cambios de schema en model_map. Si hay modelos existentes configurados, la migración necesita considerar backward compatibility.

3. **Breaking change**: Cambiar el formato de MODEL_MAP es breaking. Versión minor/major bump necesaria.

4. **Testing**: Necesita tests de integración con model_map configurado para Anthropic upstream mientras global es nim.

5. **Documentation**: El README debe actualizarse si cambiamos el formato de MODEL_MAP.

### Para Simple Fix:
- [ ] Tests con model_map apuntando a upstream Anthropic mientras global upstream_type es "nim"
- [ ] Verificar que el header `anthropic-beta` está presente en el request
- [ ] Verificar que upstreams que NO contienen "anthropic" en el nombre NO reciben el header
- [ ] `cargo test` pasa sin errores

### Para Proper Fix (si se implementa):
- [ ] Tests para model_map entries con `upstream_type: None` → usa global
- [ ] Tests para model_map entries con `upstream_type: Some("anthropic")` → usa específico
- [ ] Documentation actualizada con nuevo formato de MODEL_MAP
- [ ] Script de migración para entradas existentes
- [ ] CHANGELOG.md entry para breaking change

## Referencias

- PR #31: Dynamic context window sync between Claude Code and upstream models
- Archivos: `src/proxy/retry.rs`, `src/config.rs`
- Related: `NEXUS_UPSTREAM_TYPE` env var, `MODEL_MAP_*` configuration
- Documentación de Anthropic API: headers beta para prompt caching
- Archivo CLAUDE.md del proyecto: "anthropic-beta header conditional BY DESIGN"

### PR DESCRIPTION
**Pull request by enerBydev**
## Summary

- **Bug A**: `get_upstream_type()` ignored `upstream_name` param — now does real per-upstream lookup with global fallback
- **Bug B**: Hardcoded `anthropic-beta` header — now extracted from client headers, merged with `PROXY_MINIMUM_BETAS`, deduplicated
- **Bug C**: Outdated beta version `prompt-caching-2024-06-01` → replaced with `prompt-caching-scope-2026-01-05`
- **Bug D**: Missing `anthropic-version` header — now forwarded from client with fallback `2023-06-01`
- **Bug E**: Unconditional `chat_template_kwargs` — now only sent to NIM upstreams (Anthropic/OpenAI/OpenRouter handle thinking natively)
- **Bug F**: No startup validation — warnings for model_map routes without matching upstream config

## Key Changes

| File | Change |
|------|--------|
| `src/proxy/headers.rs` | **NEW** — `ClientHeaders` struct for header extraction/resolution |
| `src/config.rs` | Per-upstream `upstream_type` field, `UPSTREAM_<NAME>_TYPE` env var, generalized type scan, startup validation |
| `src/config_cmd.rs` | Show per-upstream type and `[type=...]` in model map entries |
| `src/proxy/mod.rs` | Extract `ClientHeaders` from request, pass through pipeline |
| `src/proxy/retry.rs` | Dynamic `anthropic-beta` + `anthropic-version` headers (conditional on Anthropic upstream) |
| `src/proxy/streaming.rs` | Thread `ClientHeaders` to `resilient_send_raw()` |
| `src/proxy/non_streaming.rs` | Thread `ClientHeaders` to `resilient_send()` |
| `src/transform.rs` | Conditional `chat_template_kwargs` (NIM only), `upstream_name` param |
| `src/transform_test.rs` | 4 new tests for conditional behavior + per-route edge case |
| `CLAUDE.md` | New env var docs, `headers.rs` module entry, updated design conventions |

## Design Decisions (from user consultation)

- **Q1 (Beta merge)**: Option C — Combine client betas + `PROXY_MINIMUM_BETAS`, deduplicate
- **Q2 (Default beta)**: Option B — Always send at least `prompt-caching-scope-2026-01-05` to Anthropic upstreams
- **Q3 (Per-route type)**: Option A — Field in `UpstreamConfig`, NOT `ModelRoute`
- **Q4 (Env var convention)**: Option A — `UPSTREAM_<NAME>_TYPE` consistent with existing pattern
- **Q5 (anthropic-version)**: Option A — Forward from client with fallback `2023-06-01`

## Test Plan

- [x] 209 tests passing (174 → +35 new)
- [x] `cargo clippy -- -D warnings` clean
- [x] `cargo fmt --check` clean
- [x] Release build OK (7.9 MB)
- [x] `cargo audit` clean
- [ ] Manual test with multi-upstream config (Anthropic + NIM)
- [ ] Verify `anthropic-beta` header present in Anthropic upstream requests
- [ ] Verify `anthropic-version` header forwarded from client
- [ ] Verify `chat_template_kwargs` absent for non-NIM upstreams
- [ ] Verify startup warnings for misconfigured model_map routes

Closes #35

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Per-upstream type overrides (env/runtime) and config display of upstream types
  * Client Anthropic header support: forwarding/merging of anthopic-beta and anthropic-version
  * Dynamic chunk timeout (env-driven) for streaming/read timeouts
  * chat_template_kwargs now applied only for NIM upstreams

* **Bug Fixes**
  * Improved streaming error handling, keep-alive/timeouts, and retry behavior

* **Tests**
  * New unit/integration tests covering header resolution, typing, and transform behavior

[Review Change Stack](https://app.coderabbit.ai/change-stack/enerBydev/nexus-ai-gateway/pull/82?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: src/proxy/headers.rs [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+    }
+
+    /// Resolve the `anthropic-version` header value to forward upstream.
+    ///
+    /// Logic (User Decision Q5, Option A):
+    /// 1. Client sends `anthropic-version` → forward it
+    /// 2. If not → use default `2023-06-01`
+    pub(crate) fn resolve_anthropic_version(&self) -> &str {
+        self.anthropic_version.as_deref().unwrap_or(DEFAULT_ANTHROPIC_VERSION)
+    }
```
[the remark this answers is not in the document]
Comment (enerBydev) [reply src/proxy/headers.rs:74]: @coderabbitai — False positive. Claude Code sends a single `anthropic-beta` header with comma-separated beta IDs (e.g. `interleaved-thinking-2025-05-14,prompt-caching-scope-2026-01-05`). The code already handles multi-beta via `split(',')` in `resolve_anthropic_beta()`. HTTP/1.1 combines multiple same-name headers into a single comma-separated value per RFC 7230 §3.2.2, so `HeaderMap::get()` is sufficient. No known client sends separate `anthropic-beta` header lines. Switching to `get_all()` would only be needed if a client violates the spec, which is a theoretical edge case not worth guarding against.

File: CLAUDE.md [hunk 2]
```diff
@@ -124,25 +125,27 @@ These behaviors are intentional and should not be changed:
 
 2. **Model identity preservation BY DESIGN** — Responses return the original Claude model ID (e.g., `claude-sonnet-4-6`) even when routed to different upstream models. This is done via `original_model` parameter in streaming responses.
 
-3. **`anthropic-beta` header conditional BY DESIGN** — Only sent to Anthropic upstream (when `NEXUS_UPSTREAM_TYPE=anthropic`). Never sent to NIM/OpenAI/OpenRouter.
+3. **`anthropic-beta` header conditional BY DESIGN** — Only sent to Anthropic upstream (when `NEXUS_UPSTREAM_TYPE=anthropic` or per-upstream `UPSTREAM_<NAME>_TYPE=anthropic`). Never sent to NIM/OpenAI/OpenRouter. Client betas are merged with `PROXY_MINIMUM_BETAS` (e.g. `prompt-caching-scope-2026-01-05`) and deduplicated. When client omits `anthropic-beta`, only proxy minimums are sent.
+
+4. **`chat_template_kwargs` conditional BY DESIGN** — `enable_thinking=true` via `chat_template_kwargs` is only included when the upstream type is NIM. Non-NIM upstreams (Anthropic, OpenAI, OpenRouter) receive the request without `chat_template_kwargs`, since they handle thinking natively.
```
[the remark this answers is not in the document]
Comment (enerBydev) [reply CLAUDE.md:130]: Fixed in commit 784f2c7 — Item 1 updated to state `chat_template_kwargs` is NIM-only, consistent with Item 4.

File: src/config.rs [hunk 3]
```diff
@@ -594,3 +716,172 @@ impl Config {
...
+    fn test_validation_warns_on_missing_type() {
+        // Model map routes to bigmodel which has no explicit type
+        let mut map = make_test_map();
+        map.insert("UPSTREAM_BIGMODEL_BASE_URL".to_string(), "http://bigmodel:11434".to_string());
+        // No UPSTREAM_BIGMODEL_TYPE set — should warn but not fail
+        map.insert("MODEL_MAP_CLAUDE_OPUS_4_6".to_string(), "bigmodel:some-model".to_string());
+        let config = Config::from_map(&map).unwrap();
+        // Config should still load successfully (warnings, not errors)
+        // Env var MODEL_MAP_CLAUDE_OPUS_4_6 → key "CLAUDE-OPUS-4-6" (underscores→hyphens)
+        assert!(config.model_map.contains_key("CLAUDE-OPUS-4-6"));
```
[the remark this answers is not in the document]
Comment (enerBydev) [reply src/config.rs]: Fixed in commit 784f2c7 — Added `.to_lowercase()` to MODEL_MAP key normalization in both `from_map()` and `from_env_with_path()`. Also fixed test assertions to use lowercase keys and corrected the invalid POSIX env var name in test.

File: src/main.rs [hunk 4]
```diff
@@ -171,7 +171,11 @@ async fn async_main(cli: Cli) -> anyhow::Result<()> {
     }
 
     let client = Client::builder()
-        .timeout(std::time::Duration::from_secs(300))
+        // CR1-fix: NO global timeout — it kills legitimate streams of 4-6 min.
+        // read_timeout is per-chunk: resets on each chunk received from upstream.
+        // Streams can run 10+ min as long as data keeps flowing.
+        // Non-streaming requests use per-request .timeout(300s) in retry.rs.
+        .read_timeout(std::time::Duration::from_secs(120))
```
[the remark this answers is not in the document]
Comment (enerBydev) [reply src/main.rs]: Fixed in commit 784f2c7 — `main.rs` now reads `chunk_timeout_secs()` instead of hardcoding 120s, so both timeouts stay aligned when `CHUNK_TIMEOUT_SECS` is customized.
